src/Saml/ResponseBuilder.php

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\GsspBundle\Saml;
+
+use SAML2_Const;
+use SAML2_Response;
+use Surfnet\SamlBundle\Exception\LogicException;
+
+final class ResponseBuilder
+{
+    private $response;
+    private $responseContext;
+
+    public function __construct(ResponseContextInterface $context)
+    {
+        if ($this->response) {
+            throw new LogicException('Cannot create a new Response when still building a response.');
+        }
+
+        $this->responseContext = $context;
+
+        $response = new SAML2_Response();
+        $response->setDestination($context->getAssertionConsumerUrl());
+        $response->setIssuer($context->getIssuer());
+        $response->setIssueInstant($context->getIssueInstant());
+        $response->setInResponseTo($context->getInResponseTo());
+
+        $this->response = $response;
+    }
+
+    /**
+     * @param string $status
+     * @param string|null $subStatus
+     * @param string|null $message
+     * @return $this
+     *
+     * @throws \Surfnet\SamlBundle\Exception\LogicException
+     */
+    public function setResponseStatus($status, $subStatus = null, $message = null)
+    {
+        if (!$this->isValidResponseStatus($status)) {
+            throw new LogicException(sprintf('Trying to set invalid Response Status'));
+        }
+
+        if ($subStatus && !$this->isValidResponseSubStatus($subStatus)) {
+            throw new LogicException(sprintf('Trying to set invalid Response SubStatus'));
+        }
+
+        $responseStatus = ['Code' => $status];
+        if ($subStatus) {
+            $responseStatus['SubCode'] = $subStatus;
+        }
+        if ($message) {
+            $responseStatus['Message'] = $message;
+        }
+
+        $this->response->setStatus($responseStatus);
+
+        return $this;
+    }
+
+    public function get()
+    {
+        $response = $this->response;
+
+        $this->response = null;
+        $this->responseContext = null;
+
+        return $response;
+    }
+
+    private function isValidResponseStatus($status)
+    {
+        return in_array($status, [
+            SAML2_Const::STATUS_SUCCESS,
+            SAML2_Const::STATUS_REQUESTER,          // Something is wrong with the AuthnRequest
+            SAML2_Const::STATUS_RESPONDER,          // Something went wrong with the Response
+            SAML2_Const::STATUS_VERSION_MISMATCH,   // The version of the request message was incorrect
+        ], true);
+    }
+
+    private function isValidResponseSubStatus($subStatus)
+    {
+        return in_array($subStatus, [
+            SAML2_Const::STATUS_AUTHN_FAILED,               // failed authentication
+            SAML2_Const::STATUS_INVALID_ATTR,
+            SAML2_Const::STATUS_INVALID_NAMEID_POLICY,
+            SAML2_Const::STATUS_NO_AUTHN_CONTEXT,           // insufficient Loa or Loa cannot be met
```

Comment:
Do you need to test for a valid Authn Context? In the case of Gateway the required Loa is loaded from the Authn Context. In this context this seems irrelevant. Please verify this with an expert ;)